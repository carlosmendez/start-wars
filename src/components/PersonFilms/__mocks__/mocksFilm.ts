export const mockFilms = [
  {
    "title": "mockFilm1",
    "releaseDate": "mockReleaseDate1",
    "planetConnection": {
      "planets": [
        {
          "surfaceWater": 1
        },
        {
          "surfaceWater": 40
        },
        {
          "surfaceWater": 8
        }
      ]
    },
    "producers": [
      "Gary Kurtz",
      "Rick McCallum"
    ]
  },
  {
    "title": "mockFilm2",
    "releaseDate": "mockReleaseDate2",
    "planetConnection": {
      "planets": [
        {
          "surfaceWater": 100
        },
        {
          "surfaceWater": 8
        },
        {
          "surfaceWater": 0
        },
        {
          "surfaceWater": 10
        }
      ]
    },
    "producers": [
      "Gary Kurtz",
      "Rick McCallum"
    ]
  },
  {
    "title": "mockFilm3",
    "releaseDate": "mockReleaseDate3",
    "planetConnection": {
      "planets": [
        {
          "surfaceWater": 1
        },
        {
          "surfaceWater": 8
        },
        {
          "surfaceWater": 8
        },
        {
          "surfaceWater": 12
        },
        {
          "surfaceWater": null
        }
      ]
    },
    "producers": [
      "Howard G. Kazanjian",
      "George Lucas",
      "Rick McCallum"
    ]
  },
  {
    "title": "4",
    "releaseDate": "mockReleaseDate4",
    "planetConnection": {
      "planets": [
        {
          "surfaceWater": 1
        },
        {
          "surfaceWater": 40
        },
        {
          "surfaceWater": 8
        },
        {
          "surfaceWater": 12
        },
        {
          "surfaceWater": null
        },
        {
          "surfaceWater": 0.9
        },
        {
          "surfaceWater": 0
        },
        {
          "surfaceWater": 60
        },
        {
          "surfaceWater": 0
        },
        {
          "surfaceWater": null
        },
        {
          "surfaceWater": null
        },
        {
          "surfaceWater": null
        },
        {
          "surfaceWater": null
        }
      ]
    },
    "producers": [
      "Rick McCallum"
    ]
  }
];
