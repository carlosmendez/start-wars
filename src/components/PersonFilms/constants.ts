export enum ButtonVisibilityEnum {
  'VISIBLE' = 'visible',
  'HIDDEN' = 'hidden',
}
