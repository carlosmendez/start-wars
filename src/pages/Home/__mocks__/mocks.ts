export const mockPeople = {
  "allPeople": {
    "edges": [
      {
        "node": {
          "name": "Luke Skywalkexxr",
          "id": "cGVvcGxlOjE=",
          "species": 50
        }
      },
      {
        "node": {
          "name": "C-3PO",
          "id": "cGVvcGxlOjI=",
          "species": {
            "averageHeight": null
          }
        }
      }
    ]
  }
};
