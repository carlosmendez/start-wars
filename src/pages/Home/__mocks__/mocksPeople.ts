export const mockPeople = {
  "allPeople": {
    "edges": [
      {
        "node": {
          "name": "mockName1",
          "id": "mockId1",
          "species": 50
        }
      },
      {
        "node": {
          "name": "mockName2",
          "id": "mockId2",
          "species": {
            "averageHeight": null
          }
        }
      }
    ]
  }
};
