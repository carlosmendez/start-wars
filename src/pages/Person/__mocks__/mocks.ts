export const mockPerson = {
  "person": {
    "name": "Luke Skywalker",
      "birthYear": "19BBY",
      "species": null,
      "filmConnection": {
      "films": [
        {
          "title": "A New Hope",
          "releaseDate": "1977-05-25",
          "planetConnection": {
            "planets": [
              {
                "surfaceWater": 1
              },
              {
                "surfaceWater": 40
              },
              {
                "surfaceWater": 8
              }
            ]
          },
          "producers": [
            "Gary Kurtz",
            "Rick McCallum"
          ]
        },
        {
          "title": "The Empire Strikes Back",
          "releaseDate": "1980-05-17",
          "planetConnection": {
            "planets": [
              {
                "surfaceWater": 100
              },
              {
                "surfaceWater": 8
              },
              {
                "surfaceWater": 0
              },
              {
                "surfaceWater": 10
              }
            ]
          },
          "producers": [
            "Gary Kurtz",
            "Rick McCallum"
          ]
        },
        {
          "title": "Return of the Jedi",
          "releaseDate": "1983-05-25",
          "planetConnection": {
            "planets": [
              {
                "surfaceWater": 1
              },
              {
                "surfaceWater": 8
              },
              {
                "surfaceWater": 8
              },
              {
                "surfaceWater": 12
              },
              {
                "surfaceWater": null
              }
            ]
          },
          "producers": [
            "Howard G. Kazanjian",
            "George Lucas",
            "Rick McCallum"
          ]
        },
        {
          "title": "Revenge of the Sith",
          "releaseDate": "2005-05-19",
          "planetConnection": {
            "planets": [
              {
                "surfaceWater": 1
              },
              {
                "surfaceWater": 40
              },
              {
                "surfaceWater": 8
              },
              {
                "surfaceWater": 12
              },
              {
                "surfaceWater": null
              },
              {
                "surfaceWater": 0.9
              },
              {
                "surfaceWater": 0
              },
              {
                "surfaceWater": 60
              },
              {
                "surfaceWater": 0
              },
              {
                "surfaceWater": null
              },
              {
                "surfaceWater": null
              },
              {
                "surfaceWater": null
              },
              {
                "surfaceWater": null
              }
            ]
          },
          "producers": [
            "Rick McCallum"
          ]
        }
      ],
        "pageInfo": {
        "hasNextPage": false,
          "hasPreviousPage": false,
          "startCursor": "YXJyYXljb25uZWN0aW9uOjA=",
          "endCursor": "YXJyYXljb25uZWN0aW9uOjM="
      },
      "totalCount": 4
    }
  }
};
