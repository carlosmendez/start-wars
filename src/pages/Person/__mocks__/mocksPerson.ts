export const mockPerson = {
  "person": {
    "name": "mockName1",
      "birthYear": "mockBirthYear",
      "species": null,
      "filmConnection": {
      "films": [
        {
          "title": "mockFilm1",
          "releaseDate": "1977-05-25",
          "planetConnection": {
            "planets": [
              {
                "surfaceWater": 1
              },
              {
                "surfaceWater": 40
              },
              {
                "surfaceWater": 8
              }
            ]
          },
          "producers": [
            "mockProducer1",
            "mockProducer2"
          ]
        },
        {
          "title": "mockFilm2",
          "releaseDate": "1980-05-17",
          "planetConnection": {
            "planets": [
              {
                "surfaceWater": 100
              },
              {
                "surfaceWater": 8
              },
              {
                "surfaceWater": 0
              },
              {
                "surfaceWater": 10
              }
            ]
          },
          "producers": [
            "mockProducer1",
            "mockProducer3"
          ]
        },
      ],
        "pageInfo": {
        "hasNextPage": false,
          "hasPreviousPage": false,
          "startCursor": "YXJyYXljb25uZWN0aW9uOjA=",
          "endCursor": "YXJyYXljb25uZWN0aW9uOjM="
      },
      "totalCount": 2
    }
  }
};
