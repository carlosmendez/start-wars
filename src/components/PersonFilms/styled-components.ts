import styled from '@emotion/styled';
import { ButtonVisibilityEnum } from "./constants";
import { COLORS, BG_COLORS } from "../../utils/ui/constants";

export const StyledFilmContainer = styled.div`
  padding: 24px;
  background-color: ${BG_COLORS.SECONDARY};
  border-radius: 8px;
  margin-top: 16px;
  max-width: 400px;
`;

export const StyleNavitationContainer = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  min-width: 350px;
  margin-top: 20px;
`;

export const StyledButton = styled.a`
  background-color: #FFF;
  border: 1px solid #000;
  border-radius: 4px;
  color: #000;
  cursor: pointer;
  padding: 10px 32px 7px;
  text-align: center;
  &:hover {
    background-color: ${COLORS.SECONDARY};
  }
  &.${ButtonVisibilityEnum.HIDDEN} {
    opacity: 0.3;
    pointer-events: none;
  }
`;
