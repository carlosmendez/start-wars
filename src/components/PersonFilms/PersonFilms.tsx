import {StyleNavitationContainer, StyledButton} from './styled-components';
import usePagination from "../../utils/hooks/usePagination";
import { ButtonVisibilityEnum } from "./constants";

type PersonFilmsProps = {
  films: FilmType[];
};

interface PlanetType {
  surfaceWater: number | null;
}

const getNumPlanetsNoWater = (planets: PlanetType[]): number =>
  planets.filter((planet: PlanetType)  => !planet.surfaceWater).length;

const PersonFilms = ({
  films,
}: PersonFilmsProps) => {
  const totalFilms = films.length;
  const {
    currentPage,
    nextPage,
    prevPage,
    hasNextPage,
    hasPrevPage,
  } = usePagination({
    total: totalFilms
  });
  const film = films[currentPage-1];
  const { title, releaseDate } = film;
  const { planetConnection } = film;
  const { planets } = planetConnection;
  const numPlanetsNoWater = getNumPlanetsNoWater(planets);
  return (
    <>
      <h2>Films</h2>
      <p>{title}</p>
      <p>{releaseDate}</p>
      <p>Number of planets without water in the film: {numPlanetsNoWater}</p>
      <StyleNavitationContainer>
        <StyledButton {...(!hasPrevPage && { visibility: ButtonVisibilityEnum.HIDDEN })} onClick={prevPage}>Prev</StyledButton>
        <StyledButton {...(!hasNextPage && { visibility: ButtonVisibilityEnum.HIDDEN })} onClick={nextPage}>Next</StyledButton>
      </StyleNavitationContainer>
    </>
  )
}

export { PersonFilms };
