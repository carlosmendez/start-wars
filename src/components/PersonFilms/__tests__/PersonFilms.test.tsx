import React from 'react';
import {render, screen} from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { PersonFilms } from "../PersonFilms";
import { mockFilms } from "../__mocks__/mocksFilm";

const props = {
  films: mockFilms
};

describe('PersonFilms component', () => {
  it('renders correctly', () => {
    const {container, getByText} = render(
      // @ts-ignore
      <PersonFilms {...props} />
    );
    expect(container).toMatchSnapshot();
    expect(getByText('mockFilm1')).toBeVisible();
    expect(getByText('Release date:')).toBeVisible();
    expect(getByText('mockReleaseDate1')).toBeVisible();
  });

  it('Should paginate Films, forward and backward', async() => {
    render(
      // @ts-ignore
      <PersonFilms {...props} />
    );
    const nextButton = screen.getByTestId('films-next');
    const prevButton = screen.getByTestId('films-prev');

    await userEvent.click(nextButton);
    expect(screen.getByText('mockFilm2')).toBeVisible();
    expect(screen.getByText('mockReleaseDate2')).toBeVisible();

    await userEvent.click(nextButton);
    expect(screen.getByText('mockFilm3')).toBeVisible();
    expect(screen.getByText('mockReleaseDate3')).toBeVisible();

    await userEvent.click(prevButton);
    expect(screen.getByText('mockFilm2')).toBeVisible();
    expect(screen.getByText('mockReleaseDate2')).toBeVisible();
  });
});
