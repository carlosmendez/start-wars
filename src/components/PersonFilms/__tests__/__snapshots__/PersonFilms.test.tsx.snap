// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PersonFilms component renders correctly 1`] = `
<div>
  <div
    class="css-8qcg57"
  >
    <h3>
      mockFilm1
    </h3>
    <p>
      <small>
        Release date: 
      </small>
      mockReleaseDate1
    </p>
    <p>
      Number of planets without water in the film: 
      0
    </p>
  </div>
  <div
    class="css-lk48ot"
  >
    <a
      class="hidden css-9w870c"
      data-testid="films-prev"
    >
      Prev
    </a>
    <a
      class=" css-9w870c"
      data-testid="films-next"
    >
      Next
    </a>
  </div>
</div>
`;
